[WIP] Better stats/ast

Rich-Harris opened the pull request:
#1891. For starters, this changes the `compiled.ast` object from this...

```js
{ html: {...}, css: [{...}], js: [{...}, {...}] }
```

...to this:

```js
{ html: {...}, css: {...}, instance: {...}, module: {...} }
```

Next up, adding `stats.vars`

Conduitry commented:
I guess I was anticipating `instance` and `module` to be keys that were still under `js`. I'm not sure which makes more sense,

Rich-Harris commented:
It could be that. I don't have a strong preference — it just seemed unnecessary maybe. 

Separately: is `stats.vars` the right name? `import` and `const` aren't vars. What would be the correct word — `symbols`?

Conduitry commented:
I suppose `vars` doesn't really bother me as much as maybe it should. `vars` seems a lot less clunky than other possibly more accurate terms like `symbols` or `identifiers`. The ESLint rule is called `no-unused-vars` and that also applies to `const`s and `import`s and `function`s and `class`es.

Also, do we need to include `'class'` as a possible identifier type in stats? Do we want to instead dial it back a bit and maybe just include whether each one is setable from the outside? (Which I guess would only apply to instance things.) I don't know where the line is with how much information to include about each thing.

Conduitry commented:
If you're soliciting feedback yet, I just tried compiling `{foo}<script></script>` and I'm getting a TypeError in `Expression.dynamic_dependencies()` at `return variable.mutated;` where `variable` is undefined.

Rich-Harris commented:
> I'm getting a TypeError in `Expression.dynamic_dependencies()`

gah. should be an easy fix, but I'm turning in for now. As well as overhauling the `stats` object, this refactors all the internals so that there's a single source of truth for everything (`component.vars`), rather than a bunch of overlapping `Set` objects describing things

Conduitry commented:
`<script>let foo; console.log($foo);</script>` results in `$foo` with `kind: 'global'`. I think this should be `kind: 'injected'`?

Rich-Harris commented:
ok I think I've addressed all the points raised above, and update the shape of `stats.vars` per earlier discussion. I like this more, it feels sensible

Conduitry commented:
As mentioned in chat, compiling `<script></script><input bind:value={foo}>` is throwing a TypeError. Under the current alpha, this compiles with only a warning - but it seems reasonable that this should instead be a compiler error, because there's really no valid way to interpret this.

Other than that, this seems good. I'm happy with how linting works with this branch. I'll cut a new version of `eslint-plugin-svelte3` once this is merged and released.

Rich-Harris commented:
released alpha21
